auth/rest_authenticator.cc

```diff
@@ -0,0 +1,440 @@
+                                                            meta::roles_valid_table::role_col_name,
+                                                            meta::roles_valid_table::qualified_name,
+                                                            meta::roles_valid_table::role_col_name);
+
+            return when_all(
+                    _qp.execute_internal(
+                            query_roles,
+                            consistency_for_user(username),
+                            internal_distributed_timeout_config(),
+                            {username},
+                            true),
+                    _qp.execute_internal(
+                            query_roles_valid,
+                            consistency_for_user(username),
+                            internal_distributed_timeout_config(),
+                            {username},
+                            true)
+            );
+        }).then_wrapped([=](future <std::tuple<future < ::shared_ptr < cql3::untyped_result_set>>,
+                            future<::shared_ptr < cql3::untyped_result_set>> >> f)
+        {
+            try {
+                auto tup = f.get0();
+                auto res_roles = std::get<0>(tup).get0();
+                auto res_roles_valid = std::get<1>(tup).get0();
+
+                auto salted_hash = std::optional<sstring>();
+                if (!res_roles->empty()) {
+                    salted_hash = res_roles->one().get_opt<sstring>(SALTED_HASH);
+                }
+                auto role_name = std::optional<sstring>();
+                if (!res_roles_valid->empty()) {
+                    role_name = res_roles_valid->one().get_opt<sstring>(meta::roles_valid_table::role_col_name);
+                }
+
+                // If not super user (super user is local only) and the username is not in roles_valid or salted_hash empty or bad password
+                // call external endpoint
+                if (username == DEFAULT_USER_NAME && (!salted_hash || !passwords::check(password, *salted_hash))) {
+                    std::throw_with_nested(exceptions::authentication_exception("Bad password for superuser"));
+                } else if (username != DEFAULT_USER_NAME &&
+                            (!role_name || !salted_hash || !passwords::check(password, *salted_hash))) {
+                    bool create_user = res_roles->empty();
+
+                    // TODO manage retry?
+                    // TODO add prometheus metrics on auth failure/success...?
+                    // TODO better delete only if date passed and repopulate async instead of ttl that just remove the entry from the table
+                    plogger.info("Authenticating username {} from rest endpoint", username);
+                    // This timeout only timebox return to client the task and callback are not cancelled
+                    return with_timeout(
```

Comment:
As you have coded it, I think timeout will apply to _rest_http_client.connect() and also its futures. To apply it only to connect, you should chain the two then() to with_timeout() instead of chaining them to _rest_http_client.connect().